File: Services/BanCheckService.cs
using Discord;
using Discord.Net;
using Discord.WebSocket;
using Microsoft.Data.Sqlite;
using ShiggyBot.Data;
using ShiggyBot.Utils;

namespace ShiggyBot.Services
{
    internal sealed class BanCheckService(DiscordSocketClient client, DatabaseService db) : IDisposable
    {
        private static readonly TimeSpan FallbackInterval = TimeSpan.FromHours(1);
        private static readonly TimeSpan Buffer = TimeSpan.FromSeconds(5);

        private readonly Lock _lock = new();
        private Timer? _timer;
        private bool _disposed;

        public void Start()
        {
            _timer = new Timer(async _ => await OnTimerAsync().ConfigureAwait(false), null, Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
            _ = ScheduleNextAsync();
            Logger.Info("[STARTUP] Ban check service started");
        }

        private async Task ScheduleNextAsync()
        {
            if (_disposed)
            {
                return;
            }

            try
            {
                DateTime? nextUnban = await db.GetNextUnbanTimeAsync().ConfigureAwait(false);

                TimeSpan delay = nextUnban.HasValue
                    ? nextUnban.Value - DateTime.UtcNow + Buffer
                    : FallbackInterval;

                if (delay < TimeSpan.Zero)
                {
                    delay = TimeSpan.Zero;
                }

                lock (_lock)
                {
                    _timer?.Change(delay, Timeout.InfiniteTimeSpan);
                }
            }
            catch (HttpRequestException ex)
            {
                ErrorHandler.LogError("Failed to query next unban time", ex);
                lock (_lock)
                {
                    _timer?.Change(FallbackInterval, Timeout.InfiniteTimeSpan);
                }
            }
            catch (SqliteException ex)
            {
                ErrorHandler.LogError("Database error scheduling next ban check", ex);
                lock (_lock)
                {
                    _timer?.Change(FallbackInterval, Timeout.InfiniteTimeSpan);
                }
            }
        }

        private async Task OnTimerAsync()
        {
            await CheckExpiredBansAsync().ConfigureAwait(false);
            _ = ScheduleNextAsync();
        }

        private async Task CheckExpiredBansAsync()
        {
            try
            {
                List<TimedBan> expiredBans = await db.GetExpiredBansAsync().ConfigureAwait(false);
                foreach (TimedBan ban in expiredBans)
                {
                    try
                    {
                        SocketGuild? guild = client.GetGuild(ban.GuildId);
                        if (guild != null)
                        {
                            await guild.RemoveBanAsync(ban.UserId).ConfigureAwait(false);
                        }

                        await db.RemoveTimedBanAsync(ban.GuildId, ban.UserId).ConfigureAwait(false);
                    }
                    catch (HttpException ex) when (ex.DiscordCode == DiscordErrorCode.UnknownBan)
                    {
                        await db.RemoveTimedBanAsync(ban.GuildId, ban.UserId).ConfigureAwait(false);
                    }
                }
            }
            catch (HttpRequestException ex)
            {
                ErrorHandler.LogError("Failed to check expired bans", ex);
            }
            catch (TaskCanceledException ex)
            {
                ErrorHandler.LogError("Timeout checking expired bans", ex);
            }
            catch (SqliteException ex)
            {
                ErrorHandler.LogError("Database error in ban check loop", ex);
            }
            catch (InvalidOperationException ex)
            {
                ErrorHandler.LogError("Discord error in ban check loop", ex);
            }
        }

        public void Stop()
        {
            lock (_lock)
            {
                _timer?.Dispose();
                _timer = null;
            }
        }

        public void Dispose()
        {
            if (_disposed)
            {
                return;
            }

            _disposed = true;
            Stop();
            GC.SuppressFinalize(this);
        }
    }
}

File: src/features/autoModeration/banCheckService.ts
import type { BotClient } from '@bot/client.js';
import { database } from '@database/index.js';
import { logger } from '@logger/index.js';

export class BanCheckService {
  private checkInterval: Timer | null = null;
  private readonly CHECK_INTERVAL_MS = 60000; // Check every minute

  async start(client: BotClient): Promise<void> {
    logger.info('Ban check service started');

    // Initial check
    await this.checkAndUnbanExpired(client);

    // Set up recurring check
    this.checkInterval = setInterval(() => {
      this.checkAndUnbanExpired(client).catch((error) => {
        logger.error('Error in ban check service', { error: error as Error });
      });
    }, this.CHECK_INTERVAL_MS);
  }

  stop(): void {
    if (this.checkInterval) {
      clearInterval(this.checkInterval);
      this.checkInterval = null;
      logger.info('Ban check service stopped');
    }
  }

  private async checkAndUnbanExpired(client: BotClient): Promise<void> {
    try {
      const expiredBans = database.getExpiredBans();

      for (const timedBan of expiredBans) {
        try {
          const guild = client.guilds.cache.get(timedBan.guildId);
          if (!guild) {
            logger.warn(`Guild ${timedBan.guildId} not found, skipping unban`);
            database.removeTimedBan(timedBan.id);
            continue;
          }

          await guild.members.unban(timedBan.userId, 'Automatic unban - timed ban expired');
          database.removeTimedBan(timedBan.id);
          logger.debug(`Unbanned user ${timedBan.userId} in guild ${timedBan.guildId}`);
        } catch (error) {
          const err = error as Error;
          logger.error(`Error unbanning user ${timedBan.userId} in guild ${timedBan.guildId}`, {
            error: err,
          });
        }
      }
    } catch (error) {
      logger.error('Error checking for expired bans', { error: error as Error });
    }
  }
}

export const banCheckService = new BanCheckService();
